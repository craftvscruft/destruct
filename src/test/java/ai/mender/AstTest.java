package ai.mender;

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

public class AstTest {

    @Test
    public void oneRuleTwoMatches() {
        CharStream inputStream = CharStreams.fromString("""
            start : { statement, moreStatements: null } => {statements}
                  | { statement, moreStatements } => {statements: [statement, ...moreStatements]};
            """
        );
        DestructListener listener = new DestructListener();
        Main.parse(inputStream, listener, true);

        Approvals.verify(Ast.toJson(listener.rules));
    }
}
