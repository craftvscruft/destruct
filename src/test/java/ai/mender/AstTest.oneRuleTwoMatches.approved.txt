[
  {
    "target": "start",
    "match": [
      {
        "left": {
          "msg": "top",
          "steps": [
            {
              "msg": "pair",
              "steps": []
            },
            {
              "msg": "take key 'statement' store to 'statement' ",
              "steps": []
            },
            {
              "msg": "pair",
              "steps": []
            },
            {
              "msg": "take key 'moreStatements' match 'null'",
              "steps": []
            }
          ]
        },
        "right": {
          "steps": []
        }
      },
      {
        "left": {
          "msg": "top",
          "steps": [
            {
              "msg": "pair",
              "steps": []
            },
            {
              "msg": "take key 'statement' store to 'statement' ",
              "steps": []
            },
            {
              "msg": "pair",
              "steps": []
            },
            {
              "msg": "take key 'moreStatements' store to 'moreStatements' ",
              "steps": []
            }
          ]
        },
        "right": {
          "steps": []
        }
      }
    ]
  }
]